react/tic-tac-toe/src/app/screens/Game/index.js

```diff
@@ -1,20 +1,108 @@
-import React from 'react';
+import React, { Component } from 'react';
 
 import styles from './styles.module.scss';
+
 import Board from './components/Board';
 
-function Game () {
-  return (
+class Game extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      history: [{
+        squares: Array(9).fill(null),
+      }],
+      stepNumber:0,
+      xIsNext: true,
+    };
+  }
+
+  handleClick(i) {
+        const history = this.state.history.slice(0, this.state.stepNumber + 1);
+    const current = history[history.length - 1];
+    const squares = current.squares.slice();
+    if (calculateWinner(squares) || squares[i]) {
+      return;
+    }
+    squares[i] = this.state.xIsNext ? 'X' : 'O';
+    this.setState({
+      history: history.concat([{
+        squares: squares,
+      }]),
+      stepNumber: history.length,
+      xIsNext: !this.state.xIsNext,
+    });
+  }
+
+  jumpTo(step) {
+    this.setState({
+      stepNumber: step,
+      xIsNext: (step % 2) === 0,
+    })
+  }
+
+  render() {
+    const history = this.state.history;
+    const current = history[this.state.stepNumber];
+    const winner = calculateWinner(current.squares);
```

Comment:
**Don't need to change this** but right now we are calculating the winner on every square click and every render.
Do you think there is a better way? What would you change to avoid the extra call to `calculateWinner`?

Reply:
I could have in the state of Game component have the winner save and only calculate when de state changes somehow.

Reply:
👏 exactly @oliferwolox , probably on each square click you would check if there is a winner, if not, then calculate it and update the state. This way we avoid calculating that on render.